README
------------
INTRODUCTION
------------

JavaScript.pm - Execute JavaScript from within Perl

JavaScript.pm is an interface to the SpiderMonkey JS engine. It lets you execute JS code, 
call JS functions, bind Perl classes to JS, import Perl functions to JS, precompile and exeute 
scripts among many other things. It does conversion between Perl and JS datatypes.

  use JavaScript;

  my $rt = JavaScript::Runtime->new();
  my $cx = $rt->create_context();

  $cx->bind_function(write => sub { print @_; });

  $cx->eval(q/
    for (i = 99; i > 0; i--) {
        write(i + " bottle(s) of beer on the wall, " + i + " bottle(s) of beer\n");
        write("Take 1 down, pass it around, ");
        if (i > 1) {
            write((i - 1) + " bottle(s) of beer on the wall.");
        }
        else {
            write("No more bottles of beer on the wall!");
        }
    }
  /);

------------
INSTALLATION
------------

To install JavaScript.pm, make sure you have libjs installed. Otherwize, grab a recent one from ftp://ftp.mozilla.org/pub/js. The module is tested with 1.5 and later.

Debian users can alternatively install the libsmjs-dev package by doing
> apt-get install libsmjs-dev

Makefile.PL will ask a number of questions regarding threading, utf8 and e4x support. Prompting for this can be disabled by 
setting their respective environment variable to a true (1) or false value (2)

    JS_THREADSAFE
    Enables or disables threadsafe. If you're getting errors in make test later that looks something like "Assertion failure: (uint32)2 < JS_MIN(((obj)->map)->freeslot, ((obj)->map)->nslots), at jsapi.c:1969" you probablly have a libjs that's threadsafe.

    JS_UTF8
    Enable utf-8 support for strings. This requires a modern spidermonkey (1.7 or later).

    JS_ENABLE_E4X
    SpiderMonkey 1.6 and later has something called E4X which is XML for JavaScript. In JavaScript these are objects (very very cool) and by enabling this we'll be able to flatten them to strings for now.

Where do look for headers and libraries can be specified using JS_LIB and JS_INC environment variables. They accept a colon (:) separated list.

The Makefile.PL will try to build a small test to determine if the libraries and headers can be used as well as checking what 
spidermonkey library is used. We require at least 1.7 but it *might* work on 1.6. If you don't want these test to be performed 
set the environment variable JS_FORCE to a true value before running Makefile.PL.

When ready invoke the following to build and install this module:

> perl Makefile.PL
> make
> make test
> make install

-------
SUPPORT
-------

= Mailing list at perl-javascript@perl.org, send an empty email to perl-javascript-subscribe@perl.org to subscribe.

Please send questions/patches/docs/comments to "Claes Jakobsson", claesjac@cpan.org

---------
COPYRIGHT
---------

Copyright (c) 2001 - 2007, Claes Jakobsson C<< <claesjac@cpan.org> >>. All rights reserved.

This module is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See http://dev.perl.org/licenses/artistic.html




Changes
Revision history for Perl extension JavaScript.

1.12  2009-04-04
	- Guard against stack corruption (Salvador Ortiz Garcia)
	- 64-bit build issues (Salvador Ortiz Garcia)
	- Updated Makefile.PL to use new META_* instead of EXTRA_META for repo.
	- Update UTF8 stuff to work with SM 1.8 (James Duncan)
	- Added a native PerlSub type that encapsulates Perl subs.
	- Converts SM regexps to Perl regexps (James Duncan)
	- Added 'apply' method to bound Perl subs (James Duncan)
    - JavaScript::Context is now a PJS_Context * (T_PTROBJ) and not a hash
    - JavaScript::Error now has a stacktrace method that returns the JS stacktrace.
    
1.11  2008-08-23
    - Added get_options, has_options and toggle_options so one can enable stuff like strict mode and JIT compilation
      if the underlying SpiderMonkey supports it (TraceMonkey ftw!).
      
1.10  2008-08-11
    - Warn user that JS_THREADSAFE is not what's recommended.
    - Added a native PerlHash type that encapsulates a Perl hash reference in JavaScript.
    - Added a native PerlArray type that encapsulates a Perl array reference in JavaScript.
    - Changed the linked list that stores bound classes to a pair of HVs storing JavaScript::PerlClass objects instead
      which are basically just PJS_Class* instances.
    - Changed the linked list that stores bound functions to an HV storing JavaScript::PerlFunction objects instead
      which are basically just PJS_Function* instances.
    - Fixed a couple of problems that cause contexts no to be destroyed when supposed to
    - Added resources section to META.yml with pointer to repository
    - Improved the test suite a bit by naming tests
    
1.09  2008-08-03
    - It's now possible to set and get the version of JS our context is dealing with since in
      some cases such as using 'let' and 'yield' we actually have to specify that we want
      1.7 features.
    - Fix a bunch of small pod issues. [rt.cpan.org #30333]
    - Fix Changes (this file) to use ISO 8601 dates
    
1.08  2008-05-08
    - Fixes Makefile.PL issues which made it fail on various platforms when trying to compile test app.
    - Added option to skip building of test app. Set environment variable JS_FORCE to a true value.
    
1.07  2008-05-06
    - Makefile.PL should abort with exit code 0 if conditions are not met instead of exit code 1.
    - Don't increase refcount for object originating from Perl (James Duncan).
    - Make bind_value croak if trying to rebind an already existing property.
    - bind_value (object|function) will now throw an error if the target already exists. [rt.cpan.org #35572]
    - Added unbind_value to remove objects, functions and value from the context. [rt.cpan.org #35572]
    - Fix longtime memory leak caused by decreasing refcount on RV instead of on the 
      SV the RV references. [rt.cpan.org #35571]
    
1.06  2008-04-18
    - Objects returned when creating constructor for bound classes should not increment
      refcount because this causes the objects to never be freed (James Duncan)
    - Docs stated argument constructor could be omitted in bind_class which wasn't the case and 
      so now it can and a default constructor will be created.
      
1.05  2008-04-01
    - Drop support for error handlers
    - This module now requires SpiderMonkey 1.7 or later.
    - Makefile.PL now compiles a small test to see if libs and headers can be found
    - Makefile.PL now discovers install locations on CentOS5 (Gavin Carr)
    - Drop support for Perl 5.6.x and earlier because we need some API symbols not available there 
      and they don't support unicode properly either. [rt.cpan.org #31114]
      
1.04  2007-11-17
    - Skip exception tests on engines less than 1.7 since these seem to be broken.
    - Free memory allocated by JS_smprintf when converting from utf8 strings (spotted by Ash Berlin)
    - Don't call js_ValueToXMLString but instead call JS method "toXMLString" on the object [rt.cpan.org #26411]
    - Convert 64bit Perl ints to JSDouble if they are too big to fit in JS ints [rt.cpan.org #26474] (Niko Tyni)
    - Add more include paths when building on debian [rt.cpan.org #26411] (Niko Tyni)
    - Fix typo in exceptions.c that didn't get global object correctly [rt.cpan.org #26411] (Niko Tyni)
    - Use JS_ExecuteScript instead of js_Execute [rt.cpan.org #26411] (Niko Tyni)
    - Allow keys/properties to be unicode (Fotango)
    - Object construction can now throw exceptions and it'll propagate to JS space [rt.cpan.org #26215]
    - Handle recursive structures better (Fotango)
    - Specify $(O_FILES) instead as a list of .o files as Writefile(OBJECT => ...)
    - Use JS_CallFunction instead of js_InternalCall so we can link aginst libjs build with hidden internals [rt.cpan.org #26411]
    - Remove support for custom stacksize since it's not what we thought it was (Fotango)
    - Add tests for exceptions (Fotango)
    - Split up giant .XS file into smaller pieces
    - Added CREDITS to MANIFEST so it's bundled with the distribution
    
1.03  2007-02-21
    - Honor value of JS_ENABLE_E4X
    - Improved README
    - Added CREDITS
    - Remove unused variables to silence some cc warnings
    - Remove ; from #include directive that can cause compilation warning (Fotango)
    - Skip t/24-c-level-interrupt-handler.t if Inline::C is not installed (Fotango)
    - Added module Test::JavaScript::More that's like Test::More but for JavaScript (Fotango)
    - Added support extending JavaScript::Runtime. This obsoletes "(void *) ext" field in PJS_Runtime and PJS_Context.
    - Added support for multple interrupt handlers. One at perl level and unlimited at C level.
    - A report->linebuf that's NULL should be undefined and not an empty string (Fotango)
    - Use SvPVutf8 macro instead of accessing via sv_2svutf8 (Fotango)
    - Added JavaScript->supports(@features) to check feature support such as e4x, ut8, threading.

Makefile.PL
use 5.008;

use strict;
use warnings;

use ExtUtils::MakeMaker;
use File::Temp qw(:POSIX);
use Config;

# CCFLAGS
my @ccflags;

push @ccflags, "-g3" if $ENV{JS_DEBUG};

# Make sure we know what platform we are compiling for
my @defines;

my $path_separator = ":";
if($^O eq "MSWin32") {
    push @defines,  "XP_PC";
    $path_separator = ";";
} else {
    push @defines, "XP_UNIX";
}

# Name of lib to link
my $lib = "js";

# Directories with include files
my @incs;

# Directories with libraries
my @libs;

# Extra header files (sans .h)
my @extra_headers;

# Directory to prefix #include with
my $inc_dir = "";

#test for debian type libsmjs
#this could probably be done better, but I'm not sure how to
#get this platform independant
if (-f "/usr/lib/libmozjs.so" or -f "/usr/local/lib/libmozjs.so") {
    $lib = "mozjs";
}
elsif (-f "/usr/lib/libsmjs.so" or -f "/usr/local/lib/libsmjs.so") {
    $lib = "smjs";
}
if (-f "/usr/include/smjs/jsapi.h") {
    push @incs, "/usr/include/smjs/";
    push @defines, "INCLUDES_IN_SMJS";
    $inc_dir = "smjs/";
}
elsif(-f "/usr/local/include/smjs/jsapi.h") {
    push @incs, "/usr/local/include/smjs/";
    push @defines, "INCLUDES_IN_SMJS";
}
elsif(-f "/usr/include/mozjs/jsapi.h") {
    push @incs, "/usr/include/mozjs/";
    push @defines, "INCLUDES_IN_MOZJS";
    $inc_dir = "mozjs/";
}
elsif(-f "/usr/local/include/mozjs/jsapi.h") {
    push @incs, "/usr/local/include/mozjs/";
    push @defines, "INCLUDES_IN_MOZJS";
}

# Debian puts nspr headers in /usr/include/nspr
if (-f "/usr/include/nspr/pratom.h") {
    push @incs, "/usr/include/nspr/";
}
# RHEL5/CentOS5 use /usr/include/nspr4 for nspr headers
elsif (-f "/usr/include/nspr4/pratom.h") {
    push @incs, "/usr/include/nspr4/";
}

# test for gentoo
if(-f "/etc/gentoo-release") {
    # Gentoo puts libjs in a slightly hidden spot.
    @incs = qw(/usr/lib/MozillaFirefox/include/js/
              /usr/lib/MozillaFirefox/);

}

# test for DarwinPorts
if (-d "/opt/local/include/spidermonkey") {
    push @incs, "/opt/local/include/spidermonkey";
    push @libs, "/opt/local/lib";
}

# Override with $ENV{JS_LIB} and $ENV{JS_INC}
if (exists $ENV{JS_LIB}) {
    @libs = get_paths($ENV{JS_LIB});
    $ENV{DYLD_LIBRARY_PATH} = $ENV{LD_LIBRARY_PATH} = $ENV{JS_LIB};
    $lib = "js";
}
if (exists $ENV{JS_INC}) {
    @incs = get_paths($ENV{JS_INC});
    @defines = grep { $_ ne "INCLUDES_IN_SMJS" } @defines;
    $inc_dir = "";
}

my $libs = join(" ", map { "-L$_" } @libs);

# Handle threadsafe
if(exists $ENV{JS_THREADSAFE}) {
    push @defines, "JS_THREADSAFE" if $ENV{JS_THREADSAFE};
}
else {
    my $enable_threadsafe = prompt("Is your SpiderMonkey compiled with JS_THREADSAFE (most things will fail if you answer wrong)? [y/N]");
    push @defines, "JS_THREADSAFE" if $enable_threadsafe eq "y";
}

if (grep /^JS_THREADSAFE$/, @defines) {
   print "Whoa there! Attention! You have asked me to build a JavaScript.pm that links against a SpiderMonkey built with ";
   print "JS_THREADSAFE defined. Although must things will proabaly work as expected some might not and I really ";
   print "don't put much effort into supporting JS_THREADSAFE (actually none expect accept some patches). ";
   print "If you are using a JS_THREADSAFE it's likely because your OS vendor supplied the SpiderMonkey libs you're using ";
   print "and instead I recommend you to download and build your own SpiderMonkey libs that doesn't use JS_THREADSAFE.\n";
}

# Check if we need to enable JS_C_STRINGS_ARE_UTF8?
if(exists $ENV{JS_UTF8}) {
    push @defines, "JS_C_STRINGS_ARE_UTF8" if $ENV{JS_UTF8};
}
else {
    my $enable_utf8 = prompt("Is your SpiderMonkey compiled with support for unicode (t/23-unicode.t will fail if you answer wrong) ? [y/N]", "N");
    push @defines, "JS_C_STRINGS_ARE_UTF8" if $enable_utf8 eq "y";
}

# Check if we want E4X support
if (exists $ENV{JS_ENABLE_E4X}) {
    if ($ENV{JS_ENABLE_E4X}) {
        push @extra_headers, "jsxml";
        push @defines, "JS_ENABLE_E4X";
    }
}
else {
    my $enable_e4x = prompt("Do you want support for E4X (requires SpiderMonkey > 1.5) ? [y/N]", "N");
    if ($enable_e4x eq "y") {
        push @extra_headers, "jsxml";
        push @defines, "JS_ENABLE_E4X";     
    }
}

# Write JavaScript_Env.h
open my $header, ">JavaScript_Env.h" || die $!;

print $header "/* This file is autogenerated to suite your platform */\n\n";

print $header "#ifndef __JAVASCRIPT_ENV_H__\n#define __JAVASCRIPT_ENV_H__\n\n";

print $header join("\n", map { "#define $_"} @defines), "\n\n";

foreach my $inc_file (qw(jsapi jsdbgapi jsinterp jsfun jsobj jsprf jsscope), @extra_headers) {
    print $header "#include <${inc_dir}${inc_file}.h>\n";
}

print $header "\n#endif\n";

close $header;

unless ($ENV{JS_FORCE}) {
    # Try a small compile to determine if we can find libs and headers
    open(my $test_script, ">", "test_js.c") || die $!;
    print $test_script <<'END_OF_SOURCE';
#include <stdio.h>
#include "JavaScript_Env.h"

int main(int argc, char **argv) {
    printf("%s", JS_GetImplementationVersion());
}
END_OF_SOURCE
    close $test_script;

    my $exe = tmpnam();;
    my $cc = join(" ", $Config{cc}, $libs, @ccflags, "-l${lib}", "-o", $exe, (map { "-I$_" } @incs), "test_js.c");
    qx($cc);
    if ($?) {
        print "Failed compiling test_js.c. ABORTING\n";
        exit 0;
    }
    unlink("test_js.c");

    # Get js version and require 1.7 or later
    my ($engine, $version, $date) = split/\s+/, qx($exe);
    my ($v2) = $version =~ /^(\d+\.\d+)/;
    if ($v2 < 1.7) {
        if (prompt("I require SpiderMonkey version 1.7 or later but found ${version}. Try anyways? [y/N]", "N") ne "y") {
            exit 0;
        }
    }

    # Dispose temp stuff
    unlink($exe);
}
else {
    print "Skipping build test since JS_FORCE is set\n";
}

# Write makefile
WriteMakefile(
    NAME            => "JavaScript",
    VERSION_FROM    => "lib/JavaScript.pm",
    PREREQ_PM       => {
        "Test::More"      => 0,
        "Test::Exception" => 0,
    },
    ABSTRACT_FROM   => "lib/JavaScript.pm", # retrieve abstract from module
    AUTHOR          => "Claes Jakobsson <claesjac\@cpan.org>",
    CCFLAGS         => join(" ", @ccflags),
    LIBS            => ["$libs -l${lib}"], # e.g., "-lm"
    INC             => join(" ", map { "-I$_" } @incs),
    LICENSE         => "perl",
    OBJECT          => q/$(O_FILES)/,
    EXTRA_META      => q{
resources:
    repository: svn://svn.versed.se/public/Perl/modules/JavaScript
},
);

sub get_paths {
    my $paths = shift;
    my @paths = map { s/^\s+//; s/\s+$//; $_ } split/$path_separator/, $paths;
    return @paths;
}

package MY;

use File::Spec;

sub post_initialize {
    my($self) = shift;

    my @headers = <*.h>;
    for (@headers, 'typemap' ) {
        $self->{PM}->{$_} = File::Spec->catfile($self->{INST_ARCHAUTODIR}, $_);
    }
    return '';
}
